### PR TITLE
upgrade to modern keras

### PR DESCRIPTION
**Pull request by sinjax**
My attempt at upgrading the version of keras used. Associated to: https://github.com/maxpumperla/elephas/issues/22

### CODE REVIEW

**Comment by maxpumperla** [comment]
Thanks a lot, I really like it. The only thing I'd rather avoid is having ```compiler``` top level in the API. This adds complexity we can avoid. Elephas already provides a compiled model on init, i.e. it comes with an ```optimizer```. So, in init of ```SparkModel``` we could do 

```{python}
self.keras_optim = master_network.optimizer.get_config()
```

```{python}
from keras.optimizers import optimizer_from_config

model = model_from_yaml(self.yaml)
keras_optimizer = optimizer_from_config(self.keras_optim)
model.compile(loss='categorical_crossentropy', optimizer=keras_optimizer, metrics=["accuracy"])
```

This way the API remains stable. Also, could we have all imports at the top? Could you please update this, then I would merge.

Thanks again, I never got around to it.

**Comment by sinjax** [comment]
Heya,

Yes I think i see, the issue is that model_from_yaml also takes a "custom_objects" parameter. This was partially the logic of exposing the loading and compilation of the model.

The primary concern is of that between the version of keras you used to use and the current one it seems that models loaded from yaml need to be told their loss function again. So it isn't correct to hardcode loss="categorical_crossentropy" for example. 

So i think the creation+compilation of the models on each worker needs to be exposed.

**Comment by maxpumperla** [comment]
Alright, the custom_objects are indeed a good point. However, I think we should go down a path that extends the current model in an obvious way. For instance, I would be fine with providing an _uncompiled_ keras model to elephas instead and providing an additional set of parameters to compile it, i.e. 

- optimizer (with default)
- loss (with default)
- custom_objects (optional)

Backward compatibility is not easy to achieve here, but it's alright for me to just focus on keras 1.x

What do you think? My gut feeling is that the current version of ```compile()``` will confuse people, because it exposes parts of the internal workings of elephas, which you shouldn't need to run such a model.

**Comment by sinjax** [comment]
I think that sounds reasonable. 

So the logic goes: give the SparkModel an uncompiled model along with the bits needed to compile it. Internally we then hold onto those bits and use them at each worker. I'll get a version done that follows that logic.

**Comment by maxpumperla** [comment]
Perfect, looking forward to it. If the SparkML stuff is too much work, don't worry, I can take up on this as well.

**Comment by sinjax** [comment]
Works with SparkModel and SparkMLModel my side, I was unsure how to alter ElephasEstimator as it seems to rely on parameters which can be serialised. Though optimizers, loss and metrics can be referred to with Strings and even custom_objects could be a dictionary of strings, sometimes they are functions...

**Comment by sinjax** [comment]
interesting! this JSONObject error occurs occasionally on my machine as well, you seen this before?

**Comment by maxpumperla** [comment]
This is new to me. Travis log seems to indicate it can't find the json that specifies the backend for keras. Strange, since you didn't touch the setup at all. 

Did that error just pop up after your changes today?

**Comment by sinjax** [comment]
yes it looks like it can't read a keras configuration or something? I'm not sure if these errors happen with or without the changes, difficult for me to investigate without downgrading keras, i'll have a poke around.

**Comment by maxpumperla** [comment]
I think we need to add this line here

https://github.com/fchollet/keras/blob/master/.travis.yml#L59

to the .travis.yml. When installing keras with pip nowadays, the ~/.keras folder is not created until you import a backend, which is why no json is found. I hope this will fix it.

**Comment by maxpumperla** [comment]
@sinjax don't worry too much about travis CI, if you can make the migration work locally, that's already awesome. I can fix CI after we merge.

**Comment by sinjax** [comment]
Been AFK last few days! Right just made the change to travis, perhaps in the right place? let's see!

**Comment by maxpumperla** [comment]
Alright, cool. Welcome back! :) You mentioned earlier this JSON error would also sometimes happen on your machine? Does that still persist?

**Comment by sinjax** [comment]
Dig dig dig. So the problem turns out to be with keras, or specifically the current released version of keras, check out this commit:
https://github.com/fchollet/keras/commit/a4d191d4f97a8100a561b87802fdd6ab5db0ed26

basically what was happening before was that regardless of whether the keras.json existed or not it was written! So in the "spark local" situation, this meant two processes were writing/reading the same file and so... non deterministic fun times!

The commit i've linked above fixes, i installed keras using their bleeding edge on github and the following command: 

`while true; do PYTHONPATH=. PYSPARK_PYTHON=python3 spark-submit --conf spark.ui.port=4000 --master=local --driver-memory 2G examples/mllib_mlp.py; done

has been merily training models for the last 30 minutes no problem where before it failed after the third repetition or so :)

**Comment by maxpumperla** [comment]
Wow, that's interesting. So that would mean in distributed mode one has to be very careful how to initialise keras on workers as well. Not so cool.

Alright, anyways: I'll merge this now. Thanks for contributing!
